### PR TITLE
[BAHIR-233] Add SNS message support for SQS streaming source

### PR DESCRIPTION
**Pull request by DmitryGrb**
Added messageWrapper option for SQS streaming connector which says if this is pure
s3 notification event or it is coming from SNS topic

### CODE REVIEW

File: sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala
Comment (abhishekd0907) [inline sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala]: Will this also throw Mapping Exception when it encounters a message in different format? Say, `S3MessageWrapper` is set to SNS but SQS queue has some message which comes directly from S3 and not via SNS

File: sql-streaming-sqs/README.md
Comment (abhishekd0907) [inline sql-streaming-sqs/README.md]: can you add a small 2-3 lines section in Readme explaining how to use S3-SQS data-source for multiple consumers  by setting `messageWrapper` to SNS? Whatever you've explained in the JIRA should be sufficient.

**Review by abhishekd0907 — COMMENTED** [review]
Overall changes LGTM. Have left a couple of comments

File: sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala
Comment (DmitryGrb) [reply sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala]: Yes, it throws MappingException, however, error message was not descriptive enough:
`org.json4s.package$MappingException: Did not find value which can be converted into java.lang.String`

So I have updated the code to explicitly check for Message field in json and if it is not there throw Mapping exception with more details. Ideally we need our own exception here, however, I used org.json4s.MappingException since it is catch a few lines later anyways and not exposed anywhere.

File: sql-streaming-sqs/README.md
Comment (DmitryGrb) [reply sql-streaming-sqs/README.md]: Done. Added small section which describes differences in SQS vs SNS approach

**Comment by abhishekd0907** [comment]
@lresende Can you please review this PR?

File: sql-streaming-sqs/README.md
Comment (abhishekd0907) [inline sql-streaming-sqs/README.md]: Maybe change the title to Using Multiple Consumers

Comment (DmitryGrb) [reply sql-streaming-sqs/README.md]: makes total sense. Changed, thanks

File: sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala
Comment (abhishekd0907) [reply sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala]: I see. There's a retry on failure mechanism for error in fetching messages from SQS. Maybe we should have an option for retry on failure in case of an error in parsing messages too. Of course, it's not related to this and can be done as a part of a separate PR.

Comment (DmitryGrb) [reply sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala]: Right now if we fail to parse SQS we just remove these messages from SQS. I think what would make sense is try to refetch them again. And agree, this should be done in another PR. I can raise a ticket and make these changes. Thoughts/Objections?

Comment (abhishekd0907) [reply sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala]: In my use case, I've seen some extra messages in the SQS queue in the following scenarios:
- When S3 bucket is configured to SQS queue during pipeline setup
- some configuration is changed 
- some test messages. 

There are not `ObjectCreate` or `ObjectRemoved ` messages and so they are not relevant for spark consumer application so it seemed reasonable to ignore and delete them. 

But if the spark consumer receives too many unparsable messages, it means there might be an issue with the pipeline set up. so according to me, it makes sense to throw an error and abort the pipeline with the help of a finite max retries mechanism.

Refetching and trying to parse the same message again may help if a legitimate message gets corrupted while fetching it from SQS. Can you list the scenarios where refetching and trying to parse the same message will help?

Feel free to raise a ticket and start a PR. We can discuss this further over there.

**Comment by abhishekd0907** [comment]
Let's wait for a review from @lresende. Is it possible to add a simple unit test to test the parsing logic for a message received via SNS? I acknowledge it might not be straightforward since `parseSqsMessages` module is private and there is no test right now for testing parsing logic for a message received directly from S3 -> SQS. You can consider making `parseSqsMessages` public if that helps.

File: sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala
Comment (DmitryGrb) [reply sql-streaming-sqs/src/main/scala/org/apache/spark/sql/streaming/sqs/SqsClient.scala]: Thanks for the details! I will raise a ticket shortly

**Comment by abhishekd0907** [comment]
@lresende Did you get a chance to review this PR? Changes LGTM

**Comment by abhishekd0907** [comment]
@lresende 
Any update on this PR from your side?

**Review by lresende — APPROVED** [review]

**Pull request merged**
